src/main/scala/io/iohk/ethereum/db/storage/AppStateStorage.scala

```diff
@@ -2,27 +2,36 @@ package io.iohk.ethereum.db.storage
 
 import io.iohk.ethereum.db.dataSource.DataSource
 import io.iohk.ethereum.db.storage.AppStateStorage._
+import io.iohk.ethereum.db.storage.pruning.PruneResult
+import io.iohk.ethereum.db.storage.pruning.PruningMode.PruneFn
 
 /**
   * This class is used to store app state variables
   *   Key: see AppStateStorage.Keys
   *   Value: stored string value
   */
-class AppStateStorage(val dataSource: DataSource) extends KeyValueStorage[Key, Value, AppStateStorage]{
+class AppStateStorage(val dataSource: DataSource, pruneFn: PruneFn) extends KeyValueStorage[Key, Value, AppStateStorage]{
   type T = AppStateStorage
 
   val namespace: IndexedSeq[Byte] = Namespaces.AppStateNamespace
   def keySerializer: Key => IndexedSeq[Byte] = _.name.getBytes
   def valueSerializer: String => IndexedSeq[Byte] = _.getBytes
   def valueDeserializer: IndexedSeq[Byte] => String = (valueBytes: IndexedSeq[Byte]) => new String(valueBytes.toArray)
 
-  protected def apply(dataSource: DataSource): AppStateStorage = new AppStateStorage(dataSource)
+  protected def apply(dataSource: DataSource): AppStateStorage = new AppStateStorage(dataSource, pruneFn)
 
   def getBestBlockNumber(): BigInt =
     BigInt(get(Keys.BestBlockNumber).getOrElse("0"))
 
-  def putBestBlockNumber(bestBlockNumber: BigInt): AppStateStorage =
+  def putBestBlockNumber(bestBlockNumber: BigInt): AppStateStorage = {
+    // FIXME We need to decouple pruning from best block number storing in this fn
+    pruneFn(getLastPrunedBlock(), bestBlockNumber) match {
+      case PruneResult(_, 0) =>
```

Comment:
In Archive mode you will still be calling `prune` but it will always return 0 and I wanted to avoid to call a "store" in that case. When using Basic mode, it will eventually update the lastPrunedBlock (even if iterates over the same block tags several times)

Your comment is totally correct but as it probably will always need to delete something (because each new block has a different root hash because of miner reward), I think it won't be a problem and we won't cal that extra save (`putLastPrunedBlock`)

if you think this is confusing (i'm starting to think it is), I can change it.